### ISSUE TITLE
Codeanalysis.Renamer.RenameSymbolAsync not finding or renaming all function symbols in class

### ISSUE THREAD

**Issue by Trackerchum** [dotnet/roslyn#28726]
**Version Used**: Microsoft.CodeAnalysis v2.8.2 with NETCore.App 2.1.0

**Steps to Reproduce**:

1. Load in a solution for an external project with AdhocWorkspace
2. Attempt to rename a function within a project with Renamer.RenameSymbolAsync
3. Write changes to physical project

**Expected Behavior**:
All symbols of that function reference are renamed to the new name

**Actual Behavior**:
The function itself is renamed however there are references within the same file that remain unchanged

This is my first time reporting a bug/issue so apologies if I've left out information or not done something correctly, there is also an open stackoverflow question regarding this with some code - https://stackoverflow.com/questions/51241809/microsoft-codeanalysis-renamer-not-renaming-or-finding-all-symbols-in-class-ne

Below is my function used to rename the symbols

```
public async Task<Solution> UpdateSymbols(Solution solution, ProjectId projectId, string oldValue, string newValue)
        {
            var project = solution.GetProject(projectId);
            var compilation = await project.GetCompilationAsync();

            var symbol = compilation.GetSymbolsWithName(s => string.Compare(s, oldValue) == 0, SymbolFilter.All).FirstOrDefault();
            if (symbol != null)
            {
                solution = await Renamer.RenameSymbolAsync(solution, symbol, newValue, project.Solution.Workspace.Options);
            }

            return solution;
        }
```

Other potentially useful information - the AdhocWorkspace is populated by recursively searching folders in a project location and adding a project and all relevant files when it finds a folder with a csproj file. It then searches each project location for the csproj file and adds the references for each project based on that.

**Comment by CyrusNajmabadi** [comment]
Hi @Trackerchum, this is likely not enough information to go off of.  Can you include a sample code file that demonstrate the problem.  i.e. the code file that contains the symbol you want to rename.

**Comment by Trackerchum** [comment]
Hi @CyrusNajmabadi  , thanks for your reply and apologies for not getting back to you sooner. I've striped the class down to it's bear minimum and replicated the issue, this is below.

`
using System;

namespace DomainTest
{
    public class Entity : BaseAggregateRoot<EntityAdded>
    {
        private Entity(Guid id) : base(id)
        {
        }

        protected override void RegisterEvents()
        {
            RegisterEvent<EntityAdded>(OnApplicationEntityAdded);
        }

        private void OnApplicationEntityAdded(EntityAdded args)
        {
            
        }
    }

    public abstract class BaseAggregateRoot<TDomainEvent>
    {
        public Guid Id
        {
            get;
            protected set;
        }
        public BaseAggregateRoot(Guid id)
        {
            this.Id = id;
        }

        protected abstract void RegisterEvents();
        protected void RegisterEvent<TEvent>(Action<TEvent> eventHandler) where TEvent : class, TDomainEvent
        {
            
        }
    }

    public class EntityAdded : DomainEvent
    { }

    public class DomainEvent : IDomainEvent
    { }

    public interface IDomainEvent
    { }

}
`

It's when trying to rename "OnApplicationEntityAdded" that the unexpected behaviour occurs, the function declaration itself is renamed but not the reference within "RegisterEvents".

Let me know if you need anymore information

**Comment by CyrusNajmabadi** [comment]
Trying out the scenario within VS shows rename working properly:

![image](https://user-images.githubusercontent.com/4564579/43126811-a696dee2-8ee2-11e8-99d6-1e272266f6d5.png)

So there must be something about doing this directly using the APIs.  Can you supply your project where you use the Roslyn APIs and where this isn't working?  Thanks!

**Comment by Trackerchum** [comment]
It's a strange one for sure! This is the project I created to test it, you'll just need to add the location of the project to the Program.cs and run it through and you'll see what I mean.

[AdHocTest.zip](https://github.com/dotnet/roslyn/files/2222771/AdHocTest.zip)

I've stripped this down to the bear minimum also, in my larger project I also add the project references.

**Comment by gamert** [comment]
I get the bug too. My solution is a litttle big(About 1500 files), the Rename will fail on some method.

**Comment by CyrusNajmabadi** [comment]
@gamert You have not provided enough information to determine what the issue is.  if you have a repro of this, can you please make it available somewhere so it can be investigated?

**Comment by Zheng979189467** [comment]
Is there any solution to this problem?I had a similar problem.

**Comment by xyz8808** [comment]
Is there any solution to this problem?I had a similar problem tooooo.

**Comment by hnim** [comment]
I got similar problem too. My full code is about 30k code lines and RenameSymbolAsync doesnt rename all function calls of some function. Is there something can work around? Thanks

**Comment by LeroyK** [comment]
This error occurs when the project is referenced through another project. For example, when you have the following project references ('=>' denotes a project reference):

ProjectA => ProjectB => ProjectC

If you rename a symbol in ProjectC that is referenced in ProjectA, it will not be renamed in ProjectA. Any references to that symbol in ProjectB will be renamed though, because it has direct project reference to ProjectC.

Can someone from the Roslyn team have a look at this and fix it?

**Comment by CyrusNajmabadi** [comment]
If project A does not reference project C, how can it use the symbols from it?  Can you supply a repro?

**Comment by LeroyK** [comment]
With project system projects, the above (ProjectA => ProjectB => ProjectC) allows ProjectA to use all symbols in ProjectC, even ProjectC's nuget package references. I tried to create a repro, but it seems to work fine there. It might have something to do with solution size, since my solution where this doesn't work has 88 projects and the repro currently only has four (I tried ProjectA => B => C =>D).

I am able to circumvent the issue by adding all redirected references in code before renaming and removing them afterwards.
